### PR TITLE
Rename `ExtensionsError`

### PR DESCRIPTION
**Pull request by khieta**
## Description of changes

* Renamed `extensions::ExtensionsError` -> `extensions::ExtensionFunctionLookupError`
* Renamed `JsonDeserializationError::ExtensionsError` -> `JsonDeserializationError::FailedExtensionFunctionLookup`

This renaming matches #188 and is also more descriptive of the error's purpose.

I believe this is a breaking change because the `JsonDeserializationError` type is exposed through the public `ContextJsonError` type. Please correct me if I'm wrong!

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [X] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-policy/CHANGELOG.md [hunk 1]
```diff
@@ -40,7 +40,8 @@
   policy evaluation.
 - Fix a panic in `PolicySet::link` that could occur when the function was called
   with a policy id corresponding to a static policy.
-- Renamed `cedar_policy_core::est::EstToAstError` to `cedar_policy_core::est::FromJsonError`
+- Renamed `cedar_policy_core::est::EstToAstError` to `cedar_policy_core::est::FromJsonError`.
+- Renamed `cedar_policy_core::entities::JsonDeserializationError::ExtensionsError` to `cedar_policy_core::entities::JsonDeserializationError::FailedExtensionsFunctionLookup`.
```
Comment (cdisselkoen) [inline cedar-policy/CHANGELOG.md:44]: could also mention the rename of `ExtensionsError` itself, not just the variant of `JsonDeserializationError`.

**Review by cdisselkoen — APPROVED** [review]

**Review by john-h-kastner-aws — APPROVED** [review]

**Review by shaobo-he-aws — APPROVED** [review]

File: cedar-policy/CHANGELOG.md (hunk 1 above)
Comment (khieta) [reply cedar-policy/CHANGELOG.md:44]: My rationale for not including this was that `ExtensionsError` is not exposed from `cedar-public`. Does that change anything?

Comment (cdisselkoen) [reply cedar-policy/CHANGELOG.md:44]: My bad, makes sense then.

**Comment by khieta** [comment]
Need 1 more review post merge with `main`

**Review by cdisselkoen — APPROVED** [review]

**Pull request merged**